### PR TITLE
PluginExtensions: Make context read only with a proxy instead of object freeze

### PR DESCRIPTION
**Pull request by mckn**
**What is this feature?**

This PR will change the way we ensure that the context passed to plugins in the UI extensions can not be modified. Instead of making a copy and freezing the context we will wrap the context in a proxy that prevents it from being modified.

It will throw a TypeError if you try to modify the values passed in the context. This is a non-breaking change since the plugin developer will be able to use the context the same way as before.

**Why do we need this feature?**

In this #67509 we add the panel data to the context. The panel data can potentially contain a lot of data which makes it far from optimal to copy and freeze. Instead we keep the same underlying data but wrap it in a proxy that will make it impossible to modify the underlying data.

**Who is this feature for?**

Plugin developers

**Which issue(s) does this PR fix?**:

None

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/plugins/extensions/utils.tsx [hunk 1]
```diff
@@ -110,3 +111,52 @@ export function generateExtensionId(pluginId: string, extensionConfig: PluginExt
...
+ * @param obj The object to make read only
+ * @returns A new read only object, does not modify the original object
+ */
+export function readOnlyProxy<T extends object>(obj: T): T {
+  if (!obj || typeof obj !== 'object' || isReadOnlyProxy(obj)) {
+    return obj;
+  }
+
+  const cache = new WeakMap();
+  const readonly = () => false;
```
Comment (leventebalogh) [inline public/app/features/plugins/extensions/utils.tsx:130]: **Nit:** maybe I would "read this easier" if this function was inlined (but maybe that's just me):

(I know like this it wouldn't be that DRY.)

```typescript
defineProperty: () => false,
deleteProperty: () => false,
isExtensible: () => false,
set: () => false,
```

File: public/app/features/plugins/extensions/utils.tsx [hunk 2]
```diff
@@ -110,3 +111,52 @@ export function generateExtensionId(pluginId: string, extensionConfig: PluginExt
...
+const _isProxy = Symbol('isReadOnlyProxy');
+
+/**
+ * Returns a proxy that wraps the given object in a way that makes it read only.
+ * If you try to modify the object a TypeError exception will be thrown.
+ *
+ * @param obj The object to make read only
+ * @returns A new read only object, does not modify the original object
+ */
+export function readOnlyProxy<T extends object>(obj: T): T {
```
Comment (leventebalogh) [inline public/app/features/plugins/extensions/utils.tsx:124]: **Nit:** maybe `getReadOnlyProxy()` instead?

**Review by leventebalogh — APPROVED** [review]
Nice one 👏 

A thought: maybe we can get rid of the `deepFreeze()` (as it's probably replaced by the new proxy-based approach now)?

Another thought: maybe if the `deepFreeze()` name is more "commonly understandable" for this kind of functionality, then we could just replace the logic of deepFreeze to the proxy-based one, and keep the name (so we don't mention the implementation - proxy - in the name)?

**Comment by mckn** [comment]
> Nice one 👏
> 
> A thought: maybe we can get rid of the `deepFreeze()` (as it's probably replaced by the new proxy-based approach now)?
> 
> Another thought: maybe if the `deepFreeze()` name is more "commonly understandable" for this kind of functionality, then we could just replace the logic of deepFreeze to the proxy-based one, and keep the name (so we don't mention the implementation - proxy - in the name)?

The `deepFreeze` is still being used by the registry.

**Pull request merged**